docs/proposals/scheduling/binding-priority-preemption/README.md

```diff
@@ -0,0 +1,290 @@
+If scheduler determines that a high priority binding cannot be scheduled even if lower priority bindings are preempted, it will not preempt lower priority bindings.
+
+### New API
+
+#### Priority Classes
+
+This proposal will reuse the PriorityClass API of Kubernetes. The priority class defines the mapping between the priority name and its value.
+It can have an optional description. It is an arbitrary string and is provided only as a guideline for users.
+
+Similarly, we will follow the semantics of PriorityClass in Karmada. The following example gives a PriorityClass used by the system by default.
+```yaml
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: foo
+globalDefault: true
+preemptionPolicy: PreemptLowerPriority
+value: 2000000000
+```
+
+**Important notes**: Only one PriorityClass can be marked as `globalDefault`. However, if more than one PriorityClasses exists with their `globalDefault` field set to true, the
+smallest value of such global default PriorityClasses will be used as the default priority.
+
+### API change
+
+#### ResourceBinding/ClusterResourceBinding API change
+
+This proposal are going to add two new field: `Priority` and `PreemptionPolicy` for determining the behavior for preempting.
+By default, preemption is disabled.
+
+```go
+// ResourceBindingSpec represents the expectation of ResourceBinding.
+type ResourceBindingSpec struct {
+    // The priority value. The karmada-scheduler component use this field to find the
+    // priority of the binding.
+    // The higher the value, the higher the priority.
+    // +optional
+    Priority *int32 `json:"priority,omitempty"`
+
+    // PreemptionBehavior is the Policy for preempting bindings with lower priority.
+    // One of Never, PreemptLowerPriority.
+    // Defaults to Never if unset.
+    // +optional
+    PreemptionBehavior PreemptionBehavior `json:"preemptionBehavior,omitempty"`
+
+    ... ...
+```
+
+#### PropagationPolicy/ClusterPropagationPolicy API change
+
+User requests sent to Karmada may have PriorityClassName in their `Placement`.
+Karmada resolves a PriorityClassName to its corresponding `Priority` and `PreemptionPolicy`, then populates them of the binding spec.
+This proposal are going to add a new field `PriorityClassName` for specify PriorityClass.
+
+```go
+// Placement represents the rule for select clusters.
+type Placement struct {
+    // PriorityClassName indicates bindings will use the PriorityClass to resolve it's priority and preemptionPolicy if specified.
+    // Any other name must be defined by creating a PriorityClass object with that name.
+    // If not specified, the binding will use the global default PriorityClass first.
+    // If the global default API does not exist, the binding priority will be zero, binding preemptionPolicy will be "Never".
+    // +optional
+    PriorityClassName string `json:"priorityClassName,omitempty"`
```

Comment:
1. I guess we can present a complete design in this proposal, and notes, currently we only implement xxx.
I mean we can have a field `PriorityClassSource`, the valid options could be `FederatedPriorityClass`,`KubePriorityClass`,`PodPriorityClass`, defaults to `KubePriorityClass`.
Note that, we might don't add this filed when implementing this feature at alpha maturity.

2. Another thing open for discussion is where the filed should be placed, under `spec.placement` or `spec`.

Reply:
sounds good, I know that `KubePriorityClass` should be `priorityclasses.scheduling.k8s.io`, but what's `PodPriorityClass`?

Reply:
The `PriorityClassSource` represents where to find the priority class, there will be 3 ways:
- From the FederatedPriorityClass, which Karmada might introduce in the future.
- From the Kubernetes PriorityClass
- From the [Pod template](https://github.com/kubernetes/kubernetes/blob/16e8911fdccbdb348a0b3291acf6102494494f9b/staging/src/k8s.io/api/core/v1/types.go#L3887-L3894)

So, the `PodPriorityClass` means parsing the priority from the pod template.

Reply:
so, you mean, for CRDs, we still need a [ResourceInterpreter interface](https://github.com/karmada-io/karmada/blob/master/pkg/resourceinterpreter/interpreter.go#L39) to determine priority?
> From the [Pod template](https://github.com/kubernetes/kubernetes/blob/16e8911fdccbdb348a0b3291acf6102494494f9b/staging/src/k8s.io/api/core/v1/types.go#L3887-L3894)

and this rule requires the priority class must exist in Karmada control plane. so in my opinion, it is not much different from `KubePriorityClass`, and the API becomes more difficult to understand.
I don't know if I understand correctly. if not,  please correct me.

Reply:
> and this rule requires the priority class must exist in Karmada control plane.

I'm not sure where validates the existence of PriorityClass in Kubernetes, but karmada-apiserver doesn't have such a restriction. People can apply Deployment and specify a non-existent PriorityClass.  

Note: The idea of `FederatedPriorityClass` and `PodPriorityClass` is still not mature, just like a placeholder in this proposal, for extension purpose.

Reply:
> I'm not sure where validates the existence of PriorityClass in Kubernetes

not K8s, I mean Karmada should do this. that's weird - Karmada can find the priority class name from the Pod template, but this priority class doesn't exist.

Reply:
The PriorityClass is a kind of cluster-lever configuration, it is usually managed by the infra team, and it might synced from the legacy metadata system. In other words, that configuration might not propagated by Karmada.

By the way, Kubernetes doesn't require the PriorityClass to exist. I just ran a test with it, it shows that everything works fine:
```yaml
-bash-5.0# karmadactl get deployments.apps --operation-scope=members --clusters=member1 nginx -o yaml 
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  namespace: default
  resourceVersion: "1428701"
  uid: 0215af9b-d8e8-44e1-bda4-92f574be2cc4
spec:
    # ...
    spec:
      containers:
      - image: nginx
      priorityClassName: horen  // not-exist PriorityClass
```

Reply:
@RainbowMango first, sorry for the slow response.
I mean, Karmada should should understand the priority class to apply it to ResourceBinding.
as your example shows, the deployment `nginx` did work fine, but it cannot co-work with ResourceBinding priority & preemption because the priority class `horen` doesn't exists in Karmada.
in plain words, when people set the field `PriorityClassSource` to `PodPriorityClass`, Karmada will parse the priority class name from the pod template, then find the priority class by name to apply it to ResourceBinding.